### PR TITLE
[WIP] Log delegate calls to Dispatcher that fail with TargetParameterCountE…

### PR DESCRIPTION
**Pull request by 333fred**
…xception.

Test only change. This change adds logging for delegates invoked from the dispatcher, so we can figure out what calls are causing TargetParameterCountExceptions in https://github.com/dotnet/roslyn/issues/21326, https://github.com/dotnet/roslyn/issues/24749, https://github.com/dotnet/roslyn/issues/25595, & https://github.com/dotnet/roslyn/issues/24754.

### CODE REVIEW

File: src/EditorFeatures/TestUtilities/DispatcherExtensions.cs [hunk 1]
```diff
@@ -9,26 +13,87 @@ public static class DispatcherExtensions
...
+        }
+
+        private static void Hooks_OperationStarted(object sender, DispatcherHookEventArgs e)
+        {
+            FieldInfo methodField = typeof(DispatcherOperation).GetField("_method", BindingFlags.NonPublic | BindingFlags.Instance);
+            var invokedDelegate = (Delegate)methodField.GetValue(e.Operation);
+            lock (s_lock)
             {
-                operation.Abort();
+                s_delegateInfos.Push(invokedDelegate.Method);
```
Comment (heejaechang) [inline src/EditorFeatures/TestUtilities/DispatcherExtensions.cs]: so is it by design not to pop until there is the exception thrown?

File: src/EditorFeatures/TestUtilities/DispatcherExtensions.cs [hunk 2]
```diff
@@ -9,26 +13,87 @@ public static class DispatcherExtensions
...
+                if (operation.Status != DispatcherOperationStatus.Completed)
+                {
+                    operation.Abort();
+                }
+            }
+            catch (TargetParameterCountException ex)
+            {
+                var methodInfoBuilder = new StringBuilder();
+                methodInfoBuilder.AppendLine("Caught TargetParameterCountException in DoEvents. Printing MethodInfos in reverse execution order.");
+                PropertyInfo fullNameProperty = typeof(MethodBase).GetProperty("FullName", BindingFlags.NonPublic | BindingFlags.Instance);
```
Comment (heejaechang) [inline src/EditorFeatures/TestUtilities/DispatcherExtensions.cs]: I dont think you actually meant all operation delegate but execution order of all nested operations? in that case, should you pop operation from delegateinfo once that operation is successfully executed?

Comment (333fred) [reply src/EditorFeatures/TestUtilities/DispatcherExtensions.cs]: Yeah, you're right. I'll update it to pop when the delegate is successfully executed.

**Comment by 333fred** [comment]
Changed title to WIP to prevent merging

**Comment by 333fred** [comment]
@dotnet-bot retest windows_release_unit32_prtest please

**Comment by 333fred** [comment]
@dotnet-bot retest windows_debug_unit32_prtest please

**Comment by 333fred** [comment]
Failed Run Links:
* https://ci.dot.net/job/dotnet_roslyn/job/dev15.7.x/job/windows_release_unit64_prtest/753/artifact/Binaries/Release/Logs/Roslyn.Services.Editor2.UnitTests.dll.out.log
* https://ci.dot.net/job/dotnet_roslyn/job/dev15.7.x/job/windows_debug_unit64_prtest/729/artifact/Binaries/Debug/Logs/Roslyn.Services.Editor2.UnitTests.dll.out.log

**Comment by sharwell** [comment]
Looks like this delegate: https://referencesource.microsoft.com/#WindowsBase/Base/System/Windows/Threading/DispatcherFrame.cs,98

**Comment by 333fred** [comment]
@dotnet-bot retest windows_release_unit32_prtest please

**Comment by 333fred** [comment]
@dotnet-bot retest windows_debug_unit32_prtest please

**Comment by 333fred** [comment]
@dotnet-bot test this please

**Comment by 333fred** [comment]
@dotnet-bot retest windows_debug_unit32_prtest please

**Comment by 333fred** [comment]
@dotnet-bot retest windows_debug_unit64_prtest please

**Pull request closed without merging**